File: LeetCodeAlgos/Python/symmetricTree.py
# Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def dps(self, left, right):
        if not left and not right:
            return True
        if not left or not right or left.val != right.val:
            return False
        else:
            return self.dps(left.left, right.right) and self.dps(left.right, right.left)
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:

        if not root:
            return True
        return self.dps(root.left, root.right)

# class Solution:
#     def isSymmetric(self, root: Optional[TreeNode]) -> bool:
#         first, next, check = [root], [], []
#         while first:
#             for n in first:
#                 if n.left:
#                     next.append(n.left)
#                     check.append(n.left.val)
#                 else:
#                     check.append(101)
#                 if n.right:
#                     next.append(n.right)
#                     check.append(n.right.val)
#                 else: 
#                     check.append(101)
#             if checkArr(check):
#                 first, next, check = next, [], []
#             else: return False
#         return True

# def checkArr(check):
#     x, y = 0, len(check)-1
#     while x<y:
#         if check[x] != check[y]:
#             return False
#         x+=1
#         y-=1
#     return True


s = Solution()
print(s.isSymmetric(TreeNode( 1, TreeNode( 2, TreeNode( 3 ), TreeNode( 4 )), TreeNode( 2, TreeNode( 4 ), TreeNode( 3 )))))

print(s.isSymmetric(TreeNode( 1, TreeNode( 2, None, TreeNode( 3 )), TreeNode( 2, None, TreeNode( 3)))))

File: LeetCodeAlgos/Java/SymmetricTree.java
// Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

public class SymmetricTree {
  public static void main(String[] args) {
    Solution s = new Solution();
    TreeNode tree1 = new TreeNode(1,
      new TreeNode(2,
        new TreeNode(3),
        new TreeNode(4)
      ),
      new TreeNode(2,
        new TreeNode(4),
        new TreeNode(3)
      )
    );
    testSymmetricTree(1, tree1, true, s);

  TreeNode tree2 = new TreeNode(1,
      new TreeNode(2,
        null,
        new TreeNode(3)
      ),
      new TreeNode(2,
        null,
        new TreeNode(3)
      )
    );
    testSymmetricTree(2, tree2, false, s);
  }

  static void testSymmetricTree(int testNum, TreeNode tree, boolean expected, Solution s) {
    boolean result = s.isSymmetric(tree);

    System.out.println(String.format(
      "Test %d: %b / %b",
      testNum,
      result,
      expected
    ));
    assert (result == expected);
  }
}

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
  public boolean isSymmetric(TreeNode root) {
    if (root.left == null && root.right == null) return true;
    else if (root.left == null ^ root.right == null) return false;
    else return followTree(root.left, root.right);
  }

  public boolean followTree(TreeNode left, TreeNode right) {
    // End of both tree lines
    if (left == null && right == null) return true;
    // If only one tree ends
    else if ((left == null ^ right == null) || left.val != right.val) {
        return false;
    }
    
    return followTree(left.left, right.right) && followTree(left.right, right.left);
  }
}